bigtable-hbase-1.x-parent/bigtable-hbase-tools/src/main/java/com/google/cloud/bigtable/hbase/tools/HBaseSchemaTranslator.java

```diff
@@ -0,0 +1,414 @@
+
+import com.google.bigtable.repackaged.com.google.common.base.Preconditions;
+import com.google.cloud.bigtable.hbase.BigtableConfiguration;
+import com.google.cloud.bigtable.hbase.tools.ClusterSchemaDefinition.TableSchemaDefinition;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HBaseConfiguration;
+import org.apache.hadoop.hbase.HRegionInfo;
+import org.apache.hadoop.hbase.HTableDescriptor;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Admin;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+import org.apache.htrace.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.htrace.fasterxml.jackson.databind.ObjectWriter;
+
+/**
+ * A utility to create tables in Cloud Bigtable based on the tables in an HBase cluster.
+ *
+ * <p>Execute the following command to copy the schema from HBase to Cloud Bigtable:
+ *
+ * <pre>
+ * java -jar bigtable-hbase-tools-1.14.1-SNAPSHOT-jar-with-dependencies.jar com.google.cloud.bigtable.hbase.tools.HBaseSchemaTranslator \
+ *  -Dhbase.zookeeper.quorum=$ZOOKEEPER_QUORUM \
+ *  -Dhbase.zookeeper.property.clientPort=$ZOOKEEPER_PORT \
+ *  -Dgoogle.bigtable.table.filter=$TABLE_NAME_REGEX \
+ *  -Dgoogle.bigtable.project.id=$PROJECT_ID \
+ *  -Dgoogle.bigtable.instance.id=$INSTANCE_ID
+ * </pre>
+ *
+ * <p>There are 2 ways to run this tool. If you can connect to both HBase and Cloud Bigtable, you
+ * can use the above method to create tables in Cloud Bigtable directly. However, if HBase master is
+ * in a private VPC or can't connect to internet, you can dump HBase schema in a file and create
+ * tables in Cloud Bigtable using that file.
+ *
+ * <p>Run the tool from a host that can connect to HBase. Store HBase schema in a file:
+ *
+ * <pre>
+ * java -jar bigtable-hbase-tools-1.14.1-SNAPSHOT-jar-with-dependencies.jar com.google.cloud.bigtable.hbase.tools.HBaseSchemaTranslator \
+ *  -Dhbase.zookeeper.quorum=$ZOOKEEPER_QUORUM \
+ *  -Dhbase.zookeeper.property.clientPort=$ZOOKEEPER_PORT \
+ *  -Dgoogle.bigtable.table.filter=$TABLE_NAME_REGEX \
+ *  -Dgoogle.bigtable.output.filepath=$SCHEMA_FILE_PATH
+ * </pre>
+ *
+ * <p>Copy the schema file to a host which can connect to Google Cloud. Create tables in Cloud
+ * Bigtable using the schema file:
+ *
+ * <pre>
+ * java -jar bigtable-hbase-tools-1.14.1-SNAPSHOT-jar-with-dependencies.jar com.google.cloud.bigtable.hbase.tools.HBaseSchemaTranslator \
+ *  -Dgoogle.bigtable.input.filepath=$SCHEMA_FILE_PATH \
+ *  -Dgoogle.bigtable.project.id=$PROJECT_ID \
+ *  -Dgoogle.bigtable.instance.id=$INSTANCE_ID
+ * </pre>
+ */
+public class HBaseSchemaTranslator {
```

Comment:
Do you expect people to use this as a library? if not, mark it as internal or package protected. If yes, please be more careful about which fields & methods are public

Reply:
marked as internal